False E0644 closure/generator type that references itself; new in nightly-2022-11-05

I tried this code with `cargo +nightly-2022-11-05 check`:

```rust
pub trait Fn0: Fn() -> Self::Out {
    type Out;
}

impl<F: Fn() -> ()> Fn0 for F {
    type Out = ();
}

pub fn closure_typer(_: impl Fn0) {}

fn main() {
    closure_typer(move || {});
}
```

I expected to see this happen: successful compilation

Instead, this happened:

```rust
error[E0644]: closure/generator type that references itself
  --> src/bin/cyclic-repro.rs:13:19
   |
13 |     closure_typer(move || {});
   |                   ^^^^^^^^^^ cyclic type of infinite size
   |
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           see issue #46062 <https://github.com/rust-lang/rust/issues/46062>
           for more information
```

It compiles successfully with the preceding nightly, `cargo +nightly-2022-11-04 check`, and stable 1.65.0.

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (09508489e 2022-11-04)
binary: rustc
commit-hash: 09508489efc223287731fe8abbd2a81bbf7adf8e
commit-date: 2022-11-04
host: x86_64-apple-darwin
release: 1.67.0-nightly
LLVM version: 15.0.4
```

To be clear, this is the nightly I installed with `rustup install nightly-2022-11-05`. I assume it's normal for the commit date to be a day earlier, but I haven't specifically noticed that before.

### Related

This sounds extremely similar to #97680, but that bug was reported months earlier whereas this one appeared today. My original discovery involved [async fn traits](https://docs.rs/async_fn_traits/0.1.1/async_fn_traits/trait.AsyncFn0.html) too, but I reduced it to an example which contains only a closure.

@rustbot label regression-from-stable-to-nightly A-closures

--- a/compiler/rustc_hir_typeck/src/closure.rs
+++ b/compiler/rustc_hir_typeck/src/closure.rs
@@ -10,6 +10,7 @@ use rustc_hir_analysis::astconv::AstConv;
 use rustc_infer::infer::type_variable::{TypeVariableOrigin, TypeVariableOriginKind};
 use rustc_infer::infer::LateBoundRegionConversionTime;
 use rustc_infer::infer::{InferOk, InferResult};
+use rustc_macros::{TypeFoldable, TypeVisitable};
 use rustc_middle::ty::subst::InternalSubsts;
 use rustc_middle::ty::visit::TypeVisitable;
 use rustc_middle::ty::{self, Ty};
@@ -22,7 +23,7 @@ use std::cmp;
 use std::iter;
 
 /// What signature do we *expect* the closure to have from context?
-#[derive(Debug)]
+#[derive(Debug, Clone, TypeFoldable, TypeVisitable)]
 struct ExpectedSig<'tcx> {
     /// Span that gave us this expectation, if we know that.
     cause_span: Option<Span>,
@@ -241,9 +242,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             if expected_sig.is_none()
                 && let ty::PredicateKind::Projection(proj_predicate) = bound_predicate.skip_binder()
             {
-                expected_sig = self.deduce_sig_from_projection(
+                expected_sig = self.normalize_associated_types_in(
+                    obligation.cause.span,
+                    self.deduce_sig_from_projection(
                     Some(obligation.cause.span),
-                    bound_predicate.rebind(proj_predicate),
+                        bound_predicate.rebind(proj_predicate),
+                    ),
                 );
             }
 

--- a/src/test/ui/closures/supertrait-hint-references-assoc-ty.rs
+++ b/src/test/ui/closures/supertrait-hint-references-assoc-ty.rs
@@ -1,0 +1,17 @@
+// check-pass
+
+pub trait Fn0: Fn(i32) -> Self::Out {
+    type Out;
+}
+
+impl<F: Fn(i32) -> ()> Fn0 for F {
+    type Out = ();
+}
+
+pub fn closure_typer(_: impl Fn0) {}
+
+fn main() {
+    closure_typer(move |x| {
+        let _: i64 = x.into();
+    });
+}

